shell/platform/darwin/macos/framework/Source/FlutterTextInputSemanticsObject.mm

```diff
@@ -0,0 +1,214 @@
+
+@end
+
+@implementation FlutterTextFieldCell {
+  NSTextView* _editor;
+}
+
+#pragma mark - Private
+
+- (instancetype)initWithTextField:(NSTextField*)textField fieldEditor:(NSTextView*)editor {
+  self = [super initTextCell:textField.stringValue];
+  if (self) {
+    _editor = editor;
+    [self setControlView:textField];
+    // Read only text fields are sent to the mac embedding as static
+    // texts. This text field must be editable and selectable at this
+    // point.
+    self.editable = YES;
+    self.selectable = YES;
+  }
+  return self;
+}
+
+#pragma mark - NSCell
+
+- (NSTextView*)fieldEditorForView:(NSView*)controlView {
+  return _editor;
+}
+
+@end
+
+#pragma mark - FlutterTextField
+
+@implementation FlutterTextField {
+  flutter::FlutterTextPlatformNode* _node;
+  FlutterTextInputPlugin* _plugin;
+}
+
+#pragma mark - Public
+
+- (instancetype)initWithPlatformNode:(flutter::FlutterTextPlatformNode*)node
+                         fieldEditor:(FlutterTextInputPlugin*)plugin {
+  self = [super initWithFrame:NSZeroRect];
+  if (self) {
+    _node = node;
+    _plugin = plugin;
+    [self setCell:[[FlutterTextFieldCell alloc] initWithTextField:self fieldEditor:plugin]];
+  }
+  return self;
+}
+
+- (void)updateString:(NSString*)string withSelection:(NSRange)selection {
+  NSAssert(_plugin.client == self,
+           @"Can't update FlutterTextField when it is not the first responder");
+  if (![[self stringValue] isEqualToString:string]) {
+    [self setStringValue:string];
+  }
+  if (!NSEqualRanges(_plugin.selectedRange, selection)) {
+    [_plugin setSelectedRange:selection];
+  }
+}
+
+#pragma mark - NSView
+
+- (NSRect)frame {
+  return _node->GetFrameRelativeToControllerView();
+}
+
+#pragma mark - NSAccessibilityProtocol
+
+- (void)setAccessibilityFocused:(BOOL)isFocused {
+  [super setAccessibilityFocused:isFocused];
+  ui::AXActionData data;
+  data.action = isFocused ? ax::mojom::Action::kFocus : ax::mojom::Action::kBlur;
+  _node->GetDelegate()->AccessibilityPerformAction(data);
+}
+
+#pragma mark - NSResponder
+
+- (BOOL)becomeFirstResponder {
+  if (!_plugin) {
+    return NO;
+  }
+  if (_plugin.client == self && [_plugin isFirstResponder]) {
+    // This text field is already the first responder.
+    return YES;
+  }
+  BOOL result = [super becomeFirstResponder];
+  if (result) {
+    _plugin.client = self;
+    // The default implementation of the becomeFirstResponder will change the
+    // text editing state. Need to manually set it back.
+    NSString* textValue = @(_node->GetStringAttribute(ax::mojom::StringAttribute::kValue).data());
+    int start = _node->GetIntAttribute(ax::mojom::IntAttribute::kTextSelStart);
+    int end = _node->GetIntAttribute(ax::mojom::IntAttribute::kTextSelEnd);
+    NSRange selection;
+    if (start > 0 && end > 0) {
+      selection = NSMakeRange(start, end - start);
+    } else {
+      selection = NSMakeRange([self stringValue].length, 0);
```

Comment:
Why is the default to set the cursor at the end of the string?

Reply:
That is the native behavior when accessibility focus on a textfield without any selection. The problem with calling becomeFirstResponder on a textfield manually is that it will select the entire string.

Reply:
setting this will sends the selection back to the framework through the textinputplugin

Reply:
sgtm - can you add a comment to that effect?